src/compiler/program.ts

```diff
@@ -765,11 +775,42 @@ namespace ts {
             const modifiedSourceFiles: { oldFile: SourceFile, newFile: SourceFile }[] = [];
             oldProgram.structureIsReused = StructureIsReused.Completely;
 
-            for (const oldSourceFile of oldProgram.getSourceFiles()) {
+            const oldSourceFiles = oldProgram.getSourceFiles();
+            for (const oldSourceFile of oldSourceFiles) {
                 const newSourceFile = host.getSourceFileByPath
                     ? host.getSourceFileByPath(oldSourceFile.fileName, oldSourceFile.path, options.target)
                     : host.getSourceFile(oldSourceFile.fileName, options.target);
 
+                const packageId = oldProgram.getPackageIdOfSourceFile(oldSourceFile);
+                if (packageId) sourceFileToPackageId.set(newSourceFile.path, packageId);

```

Comment:
i would move this to the end, when we are sure that we are completely reusing the program.